docs/python-env.md

```diff
@@ -0,0 +1,119 @@
+[^config]: [CONFIGURATION.md](../CONFIGURATION.md)
+[^rope-ast]: [`rope` discussion #557 (comment)](https://github.com/python-rope/rope/discussions/557#discussioncomment-4272747)
+[^rope-config]: [Rope docs: Configuration](https://rope.readthedocs.io/en/latest/configuration.html)
+[^pylsp-mypy-config]: [`pylsp-mypy-config` README: Configuration](https://github.com/python-lsp/pylsp-mypy?tab=readme-ov-file#configuration)
+
+### Conclusions
+
+Many plugins should really run inside the project's virtualenv, but not all of these come with configuration options to facilitate this (e.g. the Black and Rope plugins).
+
+So we can already see that it seems preferable to run `python-lsp-server` as a whole in the project environment.
+
+In the section below, we'll compare this approach to others, while the section after that will discuss how to make your editor/IDE use the correct environment.
+
+
+## Question 1: Where to install `python-lsp-server`, its plugins & their dependencies?
+
+### Option 1 (bad): In the global system or user Python env
+
+This is always a bad idea, so bad that many operating systems now make you add `--break-system-packages` to `pip` invocations when you try to do this. So it's not recommended in this case either.
+
+### Option 2 (better but still bad): In a global virtualenv specifically for `python-lsp-server` or your editor
+
+This is better than the approach above, but won't help with plugins that need to run in your project's environment and don't have configuration variables to facilitate this.
+
+### Option 3 (ok): In the project environment
+
+This is the first option that should work fine for all plugins.
+
+A potential drawbacks is that it is sometimes recommended to only install packages in your project's virtualenv that are specified by your project's `pyproject.toml` (or equivalent configs). Many Python dev tools come with functionality to ensure this (e.g. `poetry install --sync`, `uv sync` unless run with `--inexact`, ...). To use these together with this option, you have to add `python-lsp-server` and any plugins you want to use to your `pyproject.toml`'s dev dependencies section. This is OK but may be annoying for users who don't want to use LSP features or have their own setup.
+
+### Option 4 (best? but complicated): In a venv layered "on top" of the project environment
+
+It is possible to "overlay" a virtualenv on top of another [using `.pth` hacks](https://stackoverflow.com/a/72972198) or by just manually setting `PYTHONPATH` and `PATH` to contain both venvs.
+
+So you could have all your project's normal and dev dependencies installed in one venv, but then have an "overlay" venv in which you install `python-lsp-server` and the plugins you want to use, ideally at versions compatible with your project's main dependencies.
+
+One tool that can do all of this automatically is [uv](https://github.com/astral-sh/uv): Running
+
+```bash
+uv run --with python-lsp-server,pylsp-mypy pylsp
+```
```

Comment:
I just read in the [`uv run --with`](https://docs.astral.sh/uv/reference/cli/#uv-run--with) docs that it's not _quite_ as smart I thougt it was:

>When used in a project, these dependencies will be layered on top of the project environment in a separate, ephemeral environment. **These dependencies are allowed to conflict with those specified by the project.**

So I'll have to edit this to make it clear that this doesn't actually fulfill all the requirements of the "ideal" case.

Reply:
Done